### ISSUE TITLE
Possible GAN discriminator loss error!!!

### ISSUE THREAD

**Issue by WJohnnyW** [liutaocode/LivePortrait-Train#7]
```
if self.args.gan_multi_scale_mode:
            recon_img_pred_eye = self.dis_eye(
                (output_result * target_eye_mask).detach() * 2 - 1
            )
            recon_img_pred_mouth = self.dis_mouth(
                (output_result * target_mouth_mask).detach() * 2 - 1
            )

            gan_d_loss_global = single_scale_d_nonsaturating_loss(
                recon_img_pred_global, real_img_pred
            )
            gan_d_loss_eye = single_scale_d_nonsaturating_loss(
                recon_img_pred_eye, real_img_pred
            )
            gan_d_loss_mouth = single_scale_d_nonsaturating_loss(
                recon_img_pred_mouth, real_img_pred
            )
            gan_d_loss = gan_d_loss_global + gan_d_loss_eye + gan_d_loss_mouth
```
`gan_d_loss_eye` and `gan_d_loss_mouth` should be calculated by `real_eye_pred` 
and  `real_mouth_pred`  instead of  `real_img_pred`, I hope author @liutaocode can check if it's true.

**Comment by liutaocode** [comment]
Yes, I checked and there is indeed a mistake here. I’ll correct it later. In fact, due to my limited GPU memory and considering the slowdown caused by adding these discriminators, I removed these modules—including the discriminators related to eyes and mouth, as well as VGG—in subsequent training. I found that this did not have much impact on the final results.

**Comment by WJohnnyW** [comment]
可能存在的错误，在 train.py 的 validation_step 函数中，`source_img = batch['target_img']` 才符合重建的要求，而目前的实现中是`source_img = batch['source_img']`

**Comment by WJohnnyW** [comment]
另外我在训练过程中，频繁出现 gan, reconstruction, vgg loss 为 NAN 的情况，小规模的实验数据集是 HDTF @liutaocode 希望得到您的答复。

**Comment by liutaocode** [comment]
> validation_step

哈喽validation这部分我没check，因为我数据量比较大，没什么过拟合风险，训练的时候val是关闭的，我是隔段时间手动check的

**Comment by liutaocode** [comment]
> 另外我在训练过程中，频繁出现 gan, reconstruction, vgg loss 为 NAN 的情况，小规模的实验数据集是 HDTF [@liutaocode](https://github.com/liutaocode) 希望得到您的答复。

结合我的经验，我有一些建议供参考：
1、首先check数据，确保dataloader里面某些值没有nan，或者有很大值，建议所有数据clip或者检查一下，如果确认不是数据问题，大概率是gan或者loss的权重失衡了；
2、NaN在前期的确实容易出现。我是先关闭gan，只训练重建损失（这时候会有点模糊），等重建收敛到0.00x的时候，能重建出来，然后后面再慢慢加上gan以及vgg；
3、另外check一下各个loss权重，不要出现过大的，多任务损失，某个损失过大而且很难学习收敛的情况，就很容易崩溃，前期训练，一些损失无关紧要的损失可以关闭

**Comment by liutaocode** [comment]
> 另外我在训练过程中，频繁出现 gan, reconstruction, vgg loss 为 NAN 的情况，小规模的实验数据集是 HDTF [@liutaocode](https://github.com/liutaocode) 希望得到您的答复。

另外HDTF有点小，身份就200多个，大部分是正脸，最终效果可能不行，不过前期验证我觉得ok
